### PR TITLE
Add chat module

### PR DESCRIPTION
**Pull request by codeAesthetic**
## `Chat(user, roomId, channel)`

- `sendMessage(message)`
- `fetchMessages()`

### CODE REVIEW

File: src/modules/Chat.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    const chatDataToSend: ChatDetails = {
+      message,
+      roomId: this.roomId,
+      user: this.user,
+    };
+    this.channel.emit("sendMessage", chatDataToSend);
+  }
+
+  // updateMessages will update our react chat components messsages
+  async fetchMessages(updateMessages: any) {
```
Comment (chirgjn) [inline src/modules/Chat.ts:32]: https://zellwk.com/blog/converting-callbacks-to-promises/#writing-your-own-promise

File: src/modules/Chat.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { User } from "src/modules/User";
+import { Socket } from "socket.io-client";
+
+interface messageData {
+  message: string;
+  user: User;
+}
+
+type ChatDetails = messageData & { roomId: string };
```
Comment (chirgjn) [inline src/modules/Chat.ts:9]: `sendMessageRequest`?

File: src/modules/Chat.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    this.roomId = roomId;
+    this.channel = channel;
+  }
+
+  sendMessage(message: string) {
+    const chatDataToSend: ChatDetails = {
+      message,
+      roomId: this.roomId,
+      user: this.user,
+    };
```
Comment (chirgjn) [inline src/modules/Chat.ts:27]: Let's inline this inside the `emit` call.

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: src/modules/Chat.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  sendMessage(message: string) {
+    this.channel.emit("sendMessage", {
+      message,
+      roomId: this.roomId,
+      user: this.user,
+    });
+  }
+
+  async fetchMessages(): Promise<messageData> {
```
Comment (chirgjn) [inline src/modules/Chat.ts:27]: You can loose the `async` here without changing the functionality

File: src/modules/Chat.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  private channel: Socket;
+
+  constructor(user: User, roomId: string, channel: Socket) {
+    this.user = user;
+    this.roomId = roomId;
+    this.channel = channel;
+  }
+
+  sendMessage(message: string) {
+    this.channel.emit("sendMessage", {
```
Comment (chirgjn) [inline src/modules/Chat.ts:20]:
```suggestion
    this.channel.emit("postChatMessage", {
```

File: src/modules/Chat.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      message,
+      roomId: this.roomId,
+      user: this.user,
+    });
+  }
+
+  async fetchMessages(): Promise<messageData> {
+    this.channel.emit("requestAllMessages", {
```
Comment (chirgjn) [inline src/modules/Chat.ts:28]:
```suggestion
    this.channel.emit("requestAllChatMessages", {
```

File: src/modules/Chat.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+      roomId: this.roomId,
+      user: this.user,
```
Comment (chirgjn) [inline src/modules/Chat.ts:30]: I don't think we need to send back the user object for this the roomId should be enough.

File: src/modules/Chat.ts (hunk 4 above)
Comment (chirgjn) [inline src/modules/Chat.ts:27]:
```suggestion
  async fetchMessages(): Promise<messageData[]> {
```

File: src/modules/Chat.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import { User } from "src/modules/User";
+import { Socket } from "socket.io-client";
+
+interface messageData {
+  message: string;
+  user: User;
+}
```
Comment (chirgjn) [inline src/modules/Chat.ts:7]:
```suggestion
interface chatMessage {
  timestamp: number;
  sender: string;
  message: string;
}
```

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: src/modules/Chat.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    this.user = user;
+    this.roomId = roomId;
+    this.channel = channel;
+  }
+
+  sendMessage(message: string) {
+    this.channel.emit("sendMessage", {
+      message,
+      roomId: this.roomId,
+      user: this.user,
```
Comment (chirgjn) [inline src/modules/Chat.ts:23]: Let's just send back the `userId` or the auth token

File: src/modules/Chat.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    this.user = user;
+    this.roomId = roomId;
+    this.channel = channel;
+  }
+
+  sendMessage(message: string) {
+    this.channel.emit("postChatMessage", {
+      message,
+      roomId: this.roomId,
+      user: this.user?.getToken(),
```
Comment (chirgjn) [inline src/modules/Chat.ts:24]: We need the token for this call, without it the user shouldn't be able to post the message.

File: src/modules/Chat.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+      roomId: this.roomId,
+    });
+    return new Promise<chatMessage[]>((resolve, reject) => {
+      this.channel.on(
+        "fetchAllMessages",
+        (messages: chatMessage[], errorMessage: string) => {
+          if (errorMessage) {
+            reject(errorMessage);
+          }
+          return resolve(messages);
```
Comment (chirgjn) [inline src/modules/Chat.ts:39]: `return` is redundant here

File: src/modules/Chat.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    });
+  }
+
+  fetchMessages(): Promise<chatMessage[]> {
+    this.channel.emit("requestAllChatMessages", {
+      roomId: this.roomId,
+    });
+    return new Promise<chatMessage[]>((resolve, reject) => {
+      this.channel.on(
+        "fetchAllMessages",
```
Comment (chirgjn) [inline src/modules/Chat.ts:34]:
```suggestion
        "fetchAllChatMessages",
```

File: src/modules/Chat.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        (messages: chatMessage[], errorMessage: string) => {
```
Comment (chirgjn) [inline src/modules/Chat.ts:35]: By the time the socket responds back the user could be inside another room. We need to send back the `roomId` with the message list.

**Review by chirgjn — CHANGES_REQUESTED** [review]

File: src/modules/Chat.ts (hunk 13 above)
Comment (codeAesthetic) [reply src/modules/Chat.ts:35]: But the user will already have `roomId` when he creates `Chat` instance. Why does `fetchMessages` need to send back `roomId` ?

File: src/modules/Chat.ts (hunk 10 above)
Comment (codeAesthetic) [reply src/modules/Chat.ts:24]: Token has no connection as to how user will postMessage that will be handled by socket Instance

File: src/modules/Chat.ts (hunk 13 above)
Comment (chirgjn) [reply src/modules/Chat.ts:35]: Agreed, I read about `room`s in socket.io documentation and we should be good w/o 'roomId` here.

File: src/modules/Chat.ts (hunk 10 above)
Comment (chirgjn) [reply src/modules/Chat.ts:24]: We have to send back the token to verify the user

Comment (codeAesthetic) [reply src/modules/Chat.ts:24]: Can you add the suggested change? I still didn't get it.

Comment (chirgjn) [reply src/modules/Chat.ts:24]: user: this.user.getToken(),

**Review by chirgjn — APPROVED** [review]
LGTM

**Pull request merged**